File: problems/13/implement_queue_using_stacks.py
class Stack:
    def __init__(self):
        self.data = []

    def empty(self) -> bool:
        return len(self.data) == 0

    def push(self, x: int) -> None:
        self.data.append(x)

    def pop(self) -> int:
        return self.data.pop(-1)

    def peek(self) -> int:
        return self.data[-1]


class MyQueue:
    def __init__(self):
        self.stack = Stack()
        self.holding = Stack()

    def push(self, x: int) -> None:
        self.stack.push(x)

    def pop(self) -> int:
        self._swap(self.stack, self.holding)
        result = self.holding.pop()
        self._swap(self.holding, self.stack)

        return result

    def peek(self) -> int:
        self._swap(self.stack, self.holding)
        result = self.holding.peek()
        self._swap(self.holding, self.stack)

        return result

    def empty(self) -> bool:
        return self.stack.empty()

    def _swap(self, a, b):
        while not a.empty():
            b.push(a.pop())


def test_queue():
    q = MyQueue()
    q.push(1)
    q.push(2)
    assert q.peek() == 1
    assert q.pop() == 1
    assert q.empty() == False


def test_stack():
    s = Stack()
    s.push(1)
    s.push(2)
    s.push(3)

    assert s.empty() == False
    assert s.peek() == 3
    assert s.pop() == 3
    assert s.pop() == 2
    assert s.pop() == 1
    assert s.empty() == True


if __name__ == "__main__":
    import pytest
    pytest.main(["-xvv", __file__])

File: problems/13/implement_queue_using_stacks.go
package main

const StackSize = 500

type Stack struct {
	items [StackSize]int
	index int
}

func NewStack() *Stack {
	return &Stack{index: -1}
}

func (s *Stack) Push(val int) {
	s.index++
	s.items[s.index] = val
}

func (s *Stack) Peek() int {
	return s.items[s.index]
}

func (s *Stack) Pop() int {
	val := s.items[s.index]
	s.index--
	return val
}

func (s *Stack) IsEmpty() bool {
	return s.index == -1
}

type MyQueue struct {
	main *Stack
	aux  *Stack
}

func NewQueue() *MyQueue {
	return &MyQueue{
		main: NewStack(),
		aux:  NewStack(),
	}
}

func (m *MyQueue) Push(val int) {
	m.main.Push(val)
}

func (m *MyQueue) Peek() int {
	m.Swap(MainToAux)
	result := m.aux.Peek()
	m.Swap(AuxToMain)
	return result
}

func (m *MyQueue) Pop() int {
	m.Swap(MainToAux)
	result := m.aux.Pop()
	m.Swap(AuxToMain)
	return result
}

func (m *MyQueue) IsEmpty() bool {
	return m.main.IsEmpty()
}

const (
	MainToAux = 0
	AuxToMain = 1
)

func (m *MyQueue) Swap(direction int) {
	switch direction {
	case MainToAux:
		for !m.main.IsEmpty() {
			m.aux.Push(m.main.Pop())
		}
	case AuxToMain:
		for !m.aux.IsEmpty() {
			m.main.Push(m.aux.Pop())
		}
	}
}
